making Dank ERC-20 compatible

ferencdg opened the pull request:
Dank will now be compatible with https://github.com/dfinance-tech/ic-token/blob/main/spec.md

ferencdg commented on dank.sh:
```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
+
+dfxDir="/home/dan/.config/dfx"
```
I will move this out into a separate config file, either in this or in the next PR.

ferencdg commented on xtc/src/lib.rs:
```diff
@@ -1,3 +1,6 @@
+#![allow(warnings)]
```
we should get rid of this after Amber's improvement PR

ferencdg commented on xtc/src/utils.rs:
```diff
@@ -0,0 +1,21 @@
+    }
+}
+
+pub fn convert_nat_to_u32(num: Nat) -> Result<u32, String> {
```
this could probably be a generic function working for uX, but we would still need to abstract away calls like: to_u64_digits(), to_u32_digits() to be able to write that generic function

ambeeeeee commented on xtc/src/common_types.rs:
```diff
@@ -0,0 +1,117 @@
+                    None,
+                    from,
+                    to,
+                    Nat::from(transaction.cycles),
+                    Nat::from(transaction.fee),
+                    Operation::transfer,
+                    Int::from(transaction.timestamp),
+                    Nat::from(0),
+                );
+            }
+            TransactionKind::TransferFrom { caller, from, to } => {
+                tx_record = TxRecord::new(
+                    Some(caller),
+                    from,
+                    to,
+                    Nat::from(transaction.cycles),
+                    Nat::from(transaction.fee),
+                    Operation::transferFrom,
+                    Int::from(transaction.timestamp),
+                    Nat::from(0),
+                );
+            }
+            TransactionKind::Mint { to } => {
+                tx_record = TxRecord::new(
+                    None,
+                    to,
+                    to,
+                    Nat::from(transaction.cycles),
+                    Nat::from(transaction.fee),
+                    Operation::mint,
+                    Int::from(transaction.timestamp),
+                    Nat::from(0),
+                );
+            }
+            _ => {
+                return Err(());
+            }
+        }
+        Ok(tx_record)
+    }
```
While I'm not saying you need to change this, you can do something like this I believe. You can also do `let tx_record = match { ... }` if you don't like the huge `Ok` wrap with the `Try` to pass errors out.

```suggestion
    fn try_from(transaction: Transaction) -> Result<TxRecord, ()> {
        Ok(match transaction.kind {
            TransactionKind::Approve { from, to } => {
                TxRecord::new(
                    None,
                    from,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::approve,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                )
            }
            TransactionKind::Transfer { from, to } => {
                TxRecord::new(
                    None,
                    from,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::transfer,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                )
            }
            TransactionKind::TransferFrom { caller, from, to } => {
                TxRecord::new(
                    Some(caller),
                    from,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::transferFrom,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                )
            }
            TransactionKind::Mint { to } => {
                 TxRecord::new(
                    None,
                    to,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::mint,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                )
            }
            _ => {
                return Err(());
            }
        }?)
    }
```

ferencdg replied:
yeah, that is a good idea, I ended up slighly modifying your suggestion, as I didn't like the explicit return Err(())
```rust
    fn try_from(transaction: Transaction) -> Result<TxRecord, ()> {
        match transaction.kind {
            TransactionKind::Approve { from, to } => {
                Ok(TxRecord::new(
                    None,
                    from,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::approve,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                ))
            }
            TransactionKind::Transfer { from, to } => {
                Ok(TxRecord::new(
                    None,
                    from,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::transfer,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                ))
            }
            TransactionKind::TransferFrom { caller, from, to } => {
                Ok(TxRecord::new(
                    Some(caller),
                    from,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::transferFrom,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                ))
            }
            TransactionKind::Mint { to } => {
                 Ok(TxRecord::new(
                    None,
                    to,
                    to,
                    Nat::from(transaction.cycles),
                    Nat::from(transaction.fee),
                    Operation::mint,
                    Int::from(transaction.timestamp),
                    Nat::from(0),
                ))
            }
            _ => {
                Err(())
            }
        }
    }
```

ambeeeeee commented on xtc/src/history.rs:
```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
         self.history.len()
     }
+
+    pub fn get_history(&self) -> &History {
```
The standard for rust "getters" is to just use the member name, in every project I've been on and this idea is shared between all devs I've asked.

```suggestion
    pub fn history(&self) -> &History {
```

ambeeeeee commented on xtc/src/history.rs:
```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
 
+//////////////////// BEGIN OF ERC-20 ///////////////////////
+
+#[update]
+pub async fn getTransaction(index: Nat) -> TxRecord {
```
Feel free to change the function name, I had to make it different than the other `get_transaction` and I don't have as good of an understanding of this project's internals.

```suggestion
#[update(name = "getTransaction")]
pub async fn get_transaction_erc(index: Nat) -> TxRecord {
```

ambeeeeee commented on xtc/src/history.rs:
```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
+
+// TODO: ticking time bomb, we need to integrate the history service, as we
+// are always using the internal buffer here
+#[update]
+pub async fn getTransactions(start: Nat, limit: Nat) -> Vec<TxRecord> {
```
Same as above.

```suggestion
#[update(name = "getTransactions")]
pub async fn get_transactions(start: Nat, limit: Nat) -> Vec<TxRecord> {
```

ambeeeeee commented on xtc/src/history.rs:
```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
+pub async fn getTransactions(start: Nat, limit: Nat) -> Vec<TxRecord> {
+    let MAX_LIMIT = Nat::from(100);
+
+    if (limit > MAX_LIMIT) {
```
```suggestion
    if limit > MAX_LIMIT {
```

ambeeeeee commented on xtc/src/ledger.rs:
```diff
@@ -140,63 +193,127 @@ impl Ledger {
-    Unknown,
+#[update(name=transferFrom)]
+pub async fn transfer_from(from: Principal, to: Principal, amount: Nat) -> TxReceipt {
+    IsShutDown::guard();
+
+    let ic = get_context();
+    let caller = ic.caller();
+
+    crate::progress().await;
+
+    let ledger = ic.get_mut::<Ledger>();
+    let amount_u64: u64 =
+        utils::convert_nat_to_u64(amount).expect("transfer failed - unable to convert amount");
+    let fee = compute_fee(amount_u64);
+    ledger.transfer_from(&caller, &from, &to, amount_u64, fee)?;
+
+    let transaction = Transaction {
+        timestamp: ic.time(),
+        cycles: amount_u64,
+        fee,
+        kind: TransactionKind::TransferFrom {
+            caller: caller,
+            from: from,
+            to: to,
+        },
+    };
+
+    Ok(Nat::from(
+        ic.get_mut::<HistoryBuffer>().push(transaction.clone()),
+    ))
```
warning: use of deprecated function `ic_kit::get_context`: get_context is deprecated use ic_kit::ic::*

```suggestion
    IsShutDown::guard();

    let caller = ic::caller();

    crate::progress().await;

    let ledger = ic::get_mut::<Ledger>();
    let amount_u64: u64 =
        utils::convert_nat_to_u64(amount).expect("transfer failed - unable to convert amount");
    let fee = compute_fee(amount_u64);
    ledger.transfer_from(&caller, &from, &to, amount_u64, fee)?;

    let transaction = Transaction {
        timestamp: ic::time(),
        cycles: amount_u64,
        fee,
        kind: TransactionKind::TransferFrom {
            caller: caller,
            from: from,
            to: to,
        },
    };

    Ok(Nat::from(
        ic::get_mut::<HistoryBuffer>().push(transaction.clone()),
    ))
```

ambeeeeee commented on xtc/src/ledger.rs:
```diff
@@ -140,63 +193,127 @@ impl Ledger {
+pub async fn approve(to: Principal, amount: Nat) -> TxReceipt {
+    IsShutDown::guard();
+    use ic_cdk::export::candid;
+    let ic = get_context();
```
warning: use of deprecated function `ic_kit::get_context`: get_context is deprecated use ic_kit::ic::*

I can't make a GitHub suggestion due to how it works, otherwise I would.

ambeeeeee commented on xtc/src/ledger.rs:
```diff
@@ -140,63 +193,127 @@ impl Ledger {
+pub async fn transfer_erc20(to: Principal, amount: Nat) -> TxReceipt {
+    IsShutDown::guard();
+
     let ic = get_context();
```
warning: use of deprecated function `ic_kit::get_context`: get_context is deprecated use ic_kit::ic::*

ambeeeeee commented on xtc/src/ledger.rs:
```diff
@@ -481,3 +596,12 @@ mod tests {
+
+#[update]
+pub async fn balance(account: Option<Principal>) -> u64 {
+    let ic = get_context();
+    let caller = ic.caller();
+    crate::progress().await;
+    let ledger = ic.get::<Ledger>();
+    ledger.balance(&account.unwrap_or(caller))
```
warning: use of deprecated function `ic_kit::get_context`: get_context is deprecated use ic_kit::ic::*

```suggestion
    let caller = ic::caller();
    crate::progress().await;
    let ledger = ic::get::<Ledger>();
    ledger.balance(&account.unwrap_or(caller))
```

ferencdg commented on xtc/src/common_types.rs:
```diff
@@ -0,0 +1,117 @@
+    type Error = ();
+    fn try_from(transaction: Transaction) -> Result<TxRecord, ()> {
+        let mut tx_record: TxRecord;
+        match transaction.kind {
```
I was thinking about this yesterday night, the standard only support 4 + 1 events: approve, transfer, transferFrom, mint, burn, but XTC has many more. If any clients like plug needs those events, we will need to fit them in here. @rocky-fleek what events Plug rely on? I suppose it has a transaction history which shows all kind of extra events like canister calls...

kanej replied:
I think it is only transfer/mint/burn but **not** canister created and canister called.

qti3e replied:
when surfacing data for the ERC-20 standard you can convert a canisterCreated and canisterCalled events to *burn*.

rocky-fleek replied:
Plug needs the following methods to work as it is right now:
```
'balance': (arg_0: [] | [Principal]) => Promise<bigint>,
  'burn': (arg_0: {
    canister_id: Principal,
    amount: bigint
  }) => Promise<BurnResult>
  'transfer': (
    arg_0: {
      'to': Principal,
      'from': [] | [Principal],
      'amount': bigint,
    },
  ) => Promise<TransferResult>,```

ferencdg commented:
You might be looking an old version of the code, I updated it this morning.
The reason why those events were missing is because they are not part of
the standard. Yesterday I thought more about it and added those events

On Tue, Oct 5, 2021, 16:10 John Kane ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In xtc/src/common_types.rs
> <https://github.com/Psychedelic/dank/pull/36#discussion_r722043375>:
>
> > +    pub timestamp: Time,
> +    pub index: Nat,
> +}
> +
> +impl TxRecord {
> +    pub fn setIndex(mut self, index: Nat) -> Self {
> +        self.index = index;
> +        self
> +    }
> +}
> +
> +impl TryFrom<Transaction> for TxRecord {
> +    type Error = ();
> +    fn try_from(transaction: Transaction) -> Result<TxRecord, ()> {
> +        let mut tx_record: TxRecord;
> +        match transaction.kind {
>
> I think it is only transfer/mint/burn but *not* canister created and
> canister called.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Psychedelic/dank/pull/36#discussion_r722043375>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEXFZQDEOWGC5UGNZEKGZNTUFK6JFANCNFSM5FI3YC4Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

kanej commented on xtc/src/history.rs:
```diff
@@ -14,7 +17,7 @@ pub struct HistoryBuffer {
     fn default() -> Self {
         HistoryBuffer {
-            history: History::<Principal, IcBackend>::new(5_000, 2000),
+            history: History::<Principal, IcBackend>::new(4_300_000, 2000),
```
@qti3e are you happy with these values?

qti3e replied:
No, `4_300_00 / 2000 = 2150`, and that's a lot of chunks to write, it'll slow down a lot of transactions since we're still not using `ic_cdk::block_on/ic_kit::spwan`, also since `4_300_000 % 2000 = 0`, it means after each flush the recent data is also flushed.

I'd go with `new(504_000, 5_000)`

It keeps the chunk size less than 2MB, but also finishes the flush in 100 writes.

qti3e commented on xtc/src/ledger.rs:
```diff
@@ -36,56 +39,95 @@ impl Ledger {
         }
 
-        self.approve(allower, spender, allowance - amount);
+        if (allower == caller) {
```
redundant parentheses

qti3e commented on xtc/src/history.rs:
```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
+// TODO: ticking time bomb, we need to integrate the history service, as we
+// are always using the internal buffer here
+#[update(name = "getTransactions")]
+pub async fn get_transactions(start: Nat, limit: Nat) -> Vec<TxRecord> {
```
doesn't need to by async

ferencdg commented on xtc/src/ledger.rs:
```diff
@@ -36,56 +39,95 @@ impl Ledger {
+    ///    map.
     #[inline]
-    pub fn approve(&mut self, allower: &Principal, spender: &Principal, amount: u64) {
+    pub fn approve(
```
ignore this from now please

ferencdg commented on xtc/src/ledger.rs:
```diff
@@ -36,56 +39,95 @@ impl Ledger {
     #[inline]
     pub fn transfer_from(
         &mut self,
+        caller: &Principal,
```
ignore this function now please

qti3e commented on xtc/src/common_types.rs:
```diff
@@ -0,0 +1,142 @@
+    }
+}
+
+impl TryFrom<Transaction> for TxRecord {
```
As I can see all of the cases are handled in the match expression, so why are you using `TryFrom` instead of `From`?

ferencdg replied:
that was not always the case, in the original implementation I deliberately left out some cases and wanted to treat them as errors, I would leave the TryFrom for greater flexibility in the future

qti3e commented on xtc/src/common_types.rs:
```diff
@@ -0,0 +1,142 @@
+                Int::from(transaction.timestamp),
+                Nat::from(0),
+            )),
+            _ => Err(()),
```
Seems unreachable to me

ferencdg replied:
guess left it from the time where we did have some not handled cases...

qti3e commented on xtc/src/history.rs:
```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
+    (&events[start_usize..min(history_usize, start_usize + limit_usize)])
+        .into_iter()
+        .enumerate()
+        .filter_map(|tx_pair| match (tx_pair.1.clone()).try_into().ok() {
+            Some(tx) => Some((tx_pair.0, tx)),
+            _ => None,
+        })
```
The filter is not needed if you use `From` as i suggested above. Then you can move the `enumerate` after the map as well.

qti3e commented on xtc/src/history.rs:
```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
+            _ => None,
+        })
+        .map(|tx: (usize, TxRecord)| tx.1.setIndex(Nat::from(start_usize + tx.0)))
+        .collect::<Vec<TxRecord>>()
```
The return type can be inferred, no need for the `::<Vec<TxRecord>>` part.

ferencdg replied:
yeah, that code went through some refactoring, and forgot to clean it up, thanks

qti3e reviewed (commented):
Amazing work, well done, I suggested some improvements.
